### PR TITLE
[posix] add RCP capability diag module

### PR DESCRIPTION
**Pull request by zhanglongxia**
This commit adds a RCP capability diag module for verifying the RCP's capability. This commit is an initial commit that only tested several Spinel commands.

### CODE REVIEW

File: src/posix/platform/rcp_caps_diag.cpp [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "rcp_caps_diag.hpp"
+
+#if OPENTHREAD_POSIX_CONFIG_RCP_CAPS_DIAG_ENABLE
+namespace ot {
+namespace Posix {
+#define SPINEL_ENTRY(aCategory, aCommand, aKey)                                      \
```
Comment (abtink) [inline src/posix/platform/rcp_caps_diag.cpp:34]:
```suggestion
namespace Posix {

#define SPINEL_ENTRY(aCategory, aCommand, aKey)                                      \
```

File: src/posix/platform/openthread-posix-config.h [hunk 2]
```diff
@@ -429,4 +429,13 @@
...
+/**
+ * @def OPENTHREAD_POSIX_CONFIG_RCP_CAPS_DIAG_ENABLE
+ *
+ * Define as 1 to enable RCP capability diagnostic support.
+ *
+ */
+#ifndef OPENTHREAD_POSIX_CONFIG_RCP_CAPS_DIAG_ENABLE
+#define OPENTHREAD_POSIX_CONFIG_RCP_CAPS_DIAG_ENABLE OPENTHREAD_CONFIG_DIAG_ENABLE
+#endif
 #endif // OPENTHREAD_PLATFORM_POSIX_CONFIG_H_
```
Comment (abtink) [inline src/posix/platform/openthread-posix-config.h:441]:
```suggestion
#endif

#endif // OPENTHREAD_PLATFORM_POSIX_CONFIG_H_
```

**Review by abtink — APPROVED** [review]
Thanks @zhanglongxia. Looks great. 👍

File: src/posix/platform/rcp_caps_diag.hpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    /**
+     * Processes RCP capability diagnostics commands.
+     *
+     * @param[in]   aArgs           The arguments of diagnostics command line.
+     * @param[in]   aArgsLength     The number of arguments in @p aArgs.
+     * @param[out]  aOutput         The diagnostics execution result.
+     * @param[in]   aOutputMaxLen   The output buffer size.
+     *
+     * @retval  OT_ERROR_INVALID_ARGS       The command is supported but invalid arguments provided.
+     * @retval  OT_ERROR_NONE               The command is successfully process.
```
Comment (Irving-cl) [inline src/posix/platform/rcp_caps_diag.hpp:75]:
```suggestion
     * @retval  OT_ERROR_NONE               The command is successfully processed.
```

File: src/posix/platform/rcp_caps_diag.hpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    struct SpinelEntry
+    {
+        Category                          mCategory;
+        uint32_t                          mCommand;
+        spinel_prop_key_t                 mKey;
+        RcpCapsDiag::SpinelCommandHandler mHandler;
+    };
+
+    void ProcessSpinel(void);
+    void TesSpinelCommands(Category aCategory);
```
Comment (Irving-cl) [inline src/posix/platform/rcp_caps_diag.hpp:103]:
```suggestion
    void TestSpinelCommands(Category aCategory);
```

File: src/posix/platform/rcp_caps_diag.cpp [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+exit:
+    return error;
+}
+
+void RcpCapsDiag::ProcessSpinel(void)
+{
+    for (uint8_t i = 0; i < kNumCategories; i++)
+    {
+        TesSpinelCommands(static_cast<Category>(i));
```
Comment (Irving-cl) [inline src/posix/platform/rcp_caps_diag.cpp:117]:
```suggestion
        TestSpinelCommands(static_cast<Category>(i));
```

File: src/posix/platform/rcp_caps_diag.cpp [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    }
+}
+
+void RcpCapsDiag::TesSpinelCommands(Category aCategory)
```
Comment (Irving-cl) [inline src/posix/platform/rcp_caps_diag.cpp:121]:
```suggestion
void RcpCapsDiag::TestSpinelCommands(Category aCategory)
```

**Review by Irving-cl — APPROVED** [review]
LGTM 👍 

Some small suggestions.

File: demo.exp [hunk 7]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/expect -f
```
Comment (jwhui) [inline demo.exp:1]: Is including this file intentional?

Comment (zhanglongxia) [reply demo.exp]: It is for local testing. I have removed it.

**Review by jwhui — APPROVED** [review]

**Pull request merged**